src/invariant/test/InvariantTests.cpp

```diff
@@ -566,3 +589,213 @@ TEST_CASE_VERSIONS("State archival eviction invariant", "[invariant][archival]")
+                ttlKey.ttl().keyHash,
+                ContractCodeMapEntryT(
+                    std::make_shared<LedgerEntry const>(extraEntry), ttlData,
+                    100));
+        }
+        else
+        {
+            auto [extraEntry, extraTTL] =
+                createContractDataWithTTL(PERSISTENT, 1000);
+            TTLData ttlData(extraTTL.data.ttl().liveUntilLedgerSeq, 1);
+            modifiedState.mContractDataEntries.emplace(
+                InternalContractDataMapEntry(extraEntry, ttlData));
+        }
+
+        REQUIRE_THROWS_AS(app->getInvariantManager().runStateSnapshotInvariant(
+                              getLedgerState(), modifiedState),
+                          InvariantDoesNotHold);
+    };
+
+    SECTION("Extra CONTRACT_DATA entry in cache not in BL")
+    {
+        testExtraEntryInCache(false);
+    }
+
+    SECTION("Extra CONTRACT_CODE entry in cache not in BL")
+    {
+        testExtraEntryInCache(true);
+    }
+
+    SECTION("TTL mismatch between cache and BL")
+    {
+        InMemorySorobanState modifiedState =
+            lm.getInMemorySorobanStateForTesting();
+
+        // Corrupt TTL of an entry in the cache
+        auto it = modifiedState.mContractDataEntries.begin();
+        auto const& entryData = it->get();
+        LedgerEntry entryCopy = *entryData.ledgerEntry;
+
+        TTLData wrongTTL(42, 1);
+        modifiedState.mContractDataEntries.erase(it);
+        modifiedState.mContractDataEntries.emplace(
+            InternalContractDataMapEntry(entryCopy, wrongTTL));
+
+        REQUIRE_THROWS_AS(app->getInvariantManager().runStateSnapshotInvariant(
+                              getLedgerState(), modifiedState),
+                          InvariantDoesNotHold);
+    }
+
+    SECTION("Orphan TTL in BL without Soroban entry")
+    {
+        // Add an orphan TTL directly to the BucketList without going
+        // through the normal ledger path to avoid hitting cache asserts
+        auto [phantomEntry, phantomTTL] =
+            createContractDataWithTTL(PERSISTENT, 1000);
+
+        BucketTestUtils::addLiveBatchAndUpdateSnapshot(
+            *app, lm.getLastClosedLedgerHeader().header, {phantomTTL}, {}, {});
+
+        REQUIRE_THROWS_AS(
+            app->getInvariantManager().runStateSnapshotInvariant(
+                getLedgerState(), lm.getInMemorySorobanStateForTesting()),
+            InvariantDoesNotHold);
+    }
+
+    SECTION("Live entry also in hot archive")
+    {
+        // Add one of the live entries to the hot archive - this violates
+        // the invariant that no entry can be live in both BucketLists
+        BucketTestUtils::addHotArchiveBatchAndUpdateSnapshot(
+            *app, lm.getLastClosedLedgerHeader().header, {dataEntry1}, {});
+
+        REQUIRE_THROWS_AS(
+            app->getInvariantManager().runStateSnapshotInvariant(
+                getLedgerState(), lm.getInMemorySorobanStateForTesting()),
+            InvariantDoesNotHold);
+    }
```

Comment:
Would it make sense to also verify that only contract data and code entries exist in the hot archive?